ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,496 @@
+    and it doesn't make sense in many cases to have a singleton that a user creates/initializes.
+
+## Clients
+
+`Temporalio::Client` is a namespace-specific client and `Temporalio::Client::Connection` is a namespace-independent
+connection. The former has shortcut for creating the latter while it creates itself. Here's how both may look:
+
+```ruby
+module Temporalio
+  class Client
+    # This can be splatted into the constructor if one option needs to be
+    # changed. This is a Struct and not Data because we support Ruby 3.1.
+    Options = Struct.new(...)
+
+    # Connect a client (i.e. Connection.new + Client.new)
+    def self.connect(target_host, namespace, runtime: nil, ...)
+      ...
+    end
+
+    # Create client with existing connection. This must be all keyword arguments
+    # and must match the Options struct.
+    def initialize(connection:, namespace:, ...)
+      ...
+    end
+
+    # Duped options 
+    def dup_options
+      ...
+    end
+
+    # Several properties such as connection, namespace, workflow_service, etc
+
+    # Returns workflow handle
+    def start_workflow(
+      workflow,
+      *args,
+      id:,
+      task_queue:,
+      ...
+    )
+      ...
+    end
+
+    # Returns result
+    def execute_workflow(
+      workflow,
+      *args,
+      id:,
+      task_queue:,
+      ...
+    )
+      ...
+    end
+
+    # Returns workflow handle
+    def workflow_handle(id, run_id: nil, ...)
+      ...
+    end
+
+    # Lots more
+  end
+
+  class Client
+    class Connection
+      Options = Struct.new(...)
+
+      # Perform connection
+      def initialize(target_host:, runtime: nil, ...)
+        ...
+      end
+
+      # Duped options 
+      def dup_options
+        ...
+      end
+
+      # Access to raw workflow service
+      def workflow_service
+        ...
+      end
+    end
+
+    class WorkflowHandle
+      # Wait for result
+      def result(follow_runs: true)
+        ...
+      end
+
+      # Wait for result
+      def result(follow_runs: true)
+        ...
+      end
+
+      # Lots more
+    end
+
+    class CloudOperationsClient
+      # Similar to client but for cloud ops
+    end
+  end
+end
+```
+
+* There are many things missing here, but the idea is clear.
+* We expect most users and most samples to just use `Client.connect`.
+  * 💭 Why not make users create a connection themselves instead of maintaining similar options sets across
+    `Connection.initialize` and `Client.connect`? We have found most people only need a single client per connection.
+* Users will have access to raw clients via `client.workflow_service` and similar
+* Workflow handles exist with all relevant calls.
+* Note that we use the kwargs constructor + options struct pattern so that a user can access duped options, change one,
+  and reconstruct via `Client.new(**my_new_options.to_h)`
+  * ❓Is there a better way to allow people to easily change an option? We didn't want a `with` pattern because, at
```

Comment:
One very common pattern in Ruby is to have config methods (including some initializers) accept an optional block that yields the config object for modification:

```ruby
class Client
  def initialize(**options)
    @config = Configuration.new(**options)
    yield @config if block_given?

    # ...
  end
end
```

```ruby
client = Client.new do |config|
  config.namespace = "my-namespace"
end
```

I think that would feel very natural for many Rubyists, but might be a pretty large departure from how the other SDKs are configured?

Reply:
For client connect and initialization (and some others) we are wanting keyword arguments since the vast majority are optional and we're not really wanting multiple ways to configure the client (e.g. allow both block and kwargs). In this case it's more like `Client.connect("localhost:7233", "my-namespace")` which is a shortcut for `Client.new(Connection.new("localhost:7233"), "my-namespace")`. There is a PR coming either here in a couple of hours or early tomorrow that will show this.

Reply:
Here is the intentionally-limited PR: https://github.com/temporalio/sdk-ruby/pull/155. Would love feedback there, I sprinkled some PR comments throughout.